File: DONGUK/프로그래머스_WEEK 3/정수 삼각형.java
class Solution {
    public int solution(int[][] triangle) {
        
        int[][] dp = new int[triangle.length][];
        
        for(int i=0; i<triangle.length; i++){
            dp[i] = new int[triangle[i].length];
        }
        
        dp[0][0] = triangle[0][0];
        for(int i=1; i<triangle.length; i++){
            for(int j=0; j<triangle[i].length; j++){
                if(j==0 || j == triangle[i].length-1){
                    if (j == 0){
                        dp[i][j] = triangle[i][j] + dp[i-1][j];
                    }
                    else{
                        dp[i][j] = triangle[i][j] + dp[i-1][j-1];
                    }
                }
                
                else{
                    dp[i][j] = triangle[i][j] + Math.max(dp[i-1][j-1], dp[i-1][j]);
                }
            }
        }
        
        int answer = 0;
        int N = triangle.length;
        for(int i=0; i<dp[N-1].length; i++){
            answer = Math.max(answer, dp[N-1][i]);
        }
        
        return answer;
    }
}

File: JINSEOP/프로그래머스_지난_문제/정수 삼각형.cpp
#include <string>
#include <vector>
#include <algorithm>
using namespace std;

void do_dp(vector<vector<int>>& triangle) {
    for (int i = 1; i < triangle.size(); i++) {
        for (int j = 0; j < triangle[i].size(); j++) {
            if (j == 0)
                triangle[i][j] += triangle[i - 1][j];
            else if (j == (triangle[i].size() - 1))
                triangle[i][j] += triangle[i - 1][j - 1];
            else
                triangle[i][j] = max(triangle[i][j] + triangle[i - 1][j - 1],
                                     triangle[i][j] + triangle[i-1][j]);
        }
    }
}

int find_max(vector<vector<int>>& triangle) {
    return *max_element(triangle[triangle.size() - 1].begin(), triangle[triangle.size() - 1].end());
}

int solution(vector<vector<int>> triangle) {
    int answer = 0;

    do_dp(triangle);
    answer = find_max(triangle);
    return answer;
}
